src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -844,20 +854,35 @@ private void ExecuteBlockActionsCore<TBlockStartAction, TBlockAction, TNodeActio
             // Execute stateful executable node analyzers, if any.
             if (executableNodeActions.Any())
             {
+                var nodesToAnalyzeBuilder = ArrayBuilder<TNode>.GetInstance();
+                addNodesToAnalyze(executableBlocks, nodesToAnalyzeBuilder);
+
                 if (syntaxNodeActions != null)
                 {
                     Debug.Assert(getKind != null);
 
+                    var syntaxNodesToAnalyzeBuilder = ArrayBuilder<SyntaxNode>.GetInstance(nodesToAnalyzeBuilder.Count);
+                    foreach (var syntaxNode in nodesToAnalyzeBuilder)
+                        syntaxNodesToAnalyzeBuilder.Add((syntaxNode as SyntaxNode)!);
+
                     var executableNodeActionsByKind = GetNodeActionsByKind(syntaxNodeActions);
-                    var syntaxNodesToAnalyze = (IEnumerable<SyntaxNode>)getNodesToAnalyze(executableBlocks);
-                    ExecuteSyntaxNodeActions(syntaxNodesToAnalyze, executableNodeActionsByKind, analyzer, declaredSymbol, semanticModel, getKind, diagReporter, isSupportedDiagnostic, filterSpan, isGeneratedCode, hasCodeBlockStartOrSymbolStartActions: startActions.Any(), cancellationToken);
+                    ExecuteSyntaxNodeActions(syntaxNodesToAnalyzeBuilder, executableNodeActionsByKind, analyzer, declaredSymbol, semanticModel, getKind, diagReporter, isSupportedDiagnostic, filterSpan, isGeneratedCode, hasCodeBlockStartOrSymbolStartActions: startActions.Any(), cancellationToken);
+
+                    syntaxNodesToAnalyzeBuilder.Free();
                 }
                 else if (operationActions != null)
                 {
+                    var operationsToAnalyzeBuilder = ArrayBuilder<IOperation>.GetInstance(nodesToAnalyzeBuilder.Count);
+                    foreach (var syntaxNode in nodesToAnalyzeBuilder)
+                        operationsToAnalyzeBuilder.Add((syntaxNode as IOperation)!);
```

Comment:
Why `(syntax as IOperation)!` and not `(IOperation)syntaxNode`? If this does not actually implement `IOperation` shouldn't we find out here vs. null refing later in code?